src/main/java/heronarts/lx/midi/surface/APC40Mk2.java

```diff
@@ -349,17 +413,33 @@ public void onParameterChanged(LXParameter parameter) {
         sendNoteOn(0, DEVICE_ON_OFF, this.effect.enabled.isOn() ? LED_ON : LED_OFF);
       } else {
         for (int i = 0; i < this.knobs.length; ++i) {
-          if (parameter == this.knobs[i]) {
+          LXListenableNormalizedParameter knobParam = parameterForKnob(this.knobs[i]);
+          if (parameter == knobParam) {
             double normalized = (parameter instanceof CompoundParameter) ?
               ((CompoundParameter) parameter).getBaseNormalized() :
-              this.knobs[i].getNormalized();
+                    knobParam.getNormalized();
             sendControlChange(0, DEVICE_KNOB + i, (int) (normalized * 127));
             break;
           }
         }
       }
     }
 
+    // Returns what parameter should be changed when a physical knob is turned. This
+    // is trivial except for LinkedColorParameters, whose function changes depending
+    // on the surface's mode.
+    private LXListenableNormalizedParameter parameterForKnob(LXListenableParameter knob) {
+      if (knob == null || knob instanceof LXListenableNormalizedParameter) {
+        // If the knob is unused or linked to a vanilla parameter, we're done.
+        return (LXListenableNormalizedParameter) knob;
+      }
+
+      // The knob is attached to an AggregateParameter.
+      AggregateParameter agg = (AggregateParameter) knob;
```

Comment:
Even though I can't think of a case that would be something else here... better to code this defensively with an `instanceof` check to avoid `ClassCastException` crashing the whole program.